src/engine/BMInterfaceTournament.php

```diff
@@ -0,0 +1,987 @@
+     */
+    protected function join_tournament($userId, $tournId, $buttonNameArray) {
+        try {
+            $tournament = $this->load_tournament($tournId);
+
+            // convert button names to button IDs
+            $buttonIdArray = $this->game()->retrieve_button_ids(
+                array_fill(0, count($buttonNameArray), $userId),
+                $buttonNameArray
+            );
+
+            if (!$this->validate_join_tournament($userId, $tournId, $tournament, $buttonIdArray)) {
+                return NULL;
+            }
+
+            $this->resolve_random_button_selection_tourn($buttonIdArray);
+
+            $success = $this->add_user_to_tournament($userId, $tournId, $tournament->nPlayers, $buttonIdArray);
+
+            if (!$success) {
+                // something's gone wrong between validation and attempting to add the user
+                $tournament = $this->load_tournament($tournId);
+                $validation = $this->validate_join_tournament($userId, $tournId, $tournament);
+
+                if ($validation) {
+                    error_log('Paradoxical validation success in BMInterface::join_tournament');
+                    $this->set_message('Tournament join failed even though it should have succeeded');
+                }
+                return NULL;
+            }
+
+            $this->watch_tournament($userId, $tournId);
+        } catch (BMExceptionDatabase $e) {
+            $this->set_message('Cannot join tournament because a player or tourn ID was not valid');
+            return NULL;
+        } catch (Exception $e) {
+            error_log(
+                'Caught exception in BMInterface::join_tournament: ' .
+                $e->getMessage()
+            );
+            $this->set_message("Tournament join failed: $e");
+            return NULL;
+        }
+
+        $this->update_tournament($tournId);
+        $this->set_message('You have successfully joined this tournament');
+        return TRUE;
+    }
+
+    /**
+     * Validate whether a player can join a tournament
+     *
+     * @param int $userId
+     * @param int $tournId
+     * @param BMTournament $tournament
+     * @param array $buttonIdArray
+     * @return bool
+     */
+    protected function validate_join_tournament($userId, $tournId, $tournament, $buttonIdArray) {
+        if (is_null($tournament)) {
+            $this->set_message("Tournament $tournId does not exist");
+            return NULL;
+        }
+
+        if ($tournament->tournamentState > BMTournamentState::JOIN_TOURNAMENT) {
+            $this->set_message('The tournament has already started.');
+            return NULL;
+        }
+
+        if (in_array($userId, $tournament->playerIdArray)) {
+            $this->set_message('You are already part of this tournament.');
+            return NULL;
+        }
+
+        if (!$tournament->validate_button_choice($buttonIdArray)) {
+            $this->set_message('Button not correctly selected.');
```

Comment:
Maybe "Button choice not valid"?  "correctly selected" makes me think the player clicked on a button wrong.

Reply:
Changed error message.